src/connectors/dynamic-widgets/connectDynamicWidgets.ts

```diff
@@ -0,0 +1,106 @@
+import { SearchResults } from 'algoliasearch-helper';
+import { getWidgetAttribute, noop } from '../../lib/utils';
+import { Connector, Widget } from '../../types';
+
+export type DynamicWidgetsRendererOptions = {
+  attributesToRender: string[];
+};
+
+export type DynamicWidgetsConnectorParams = {
+  widgets: Widget[];
+  transformItems?(
+    items: string[],
+    metadata: { results: SearchResults }
+  ): string[];
+};
+
+export type DynamicWidgetsConnector = Connector<
+  DynamicWidgetsRendererOptions,
+  DynamicWidgetsConnectorParams
+>;
+
+const connectDynamicWidgets: DynamicWidgetsConnector = function connectDynamicWidgets(
+  renderFn,
+  unmountFn = noop
+) {
+  return widgetParams => {
+    const { widgets, transformItems = items => items } = widgetParams;
+
+    const localWidgets: Map<
+      string,
+      { widget: Widget; isMounted: boolean }
+    > = new Map();
+
+    return {
+      $$type: 'ais.dynamicWidgets',
+      init(initOptions) {
+        widgets.forEach(widget => {
+          const attribute = getWidgetAttribute(widget, initOptions);
+          localWidgets.set(attribute, { widget, isMounted: true });
+        });
+        initOptions.parent!.addWidgets(widgets);
+      },
+      render(renderOptions) {
+        const { parent } = renderOptions;
+        const renderState = this.getWidgetRenderState(renderOptions);
+
+        const widgetsToUnmount: Widget[] = [];
+        const widgetsToMount: Widget[] = [];
+
+        localWidgets.forEach(({ widget, isMounted }, attribute) => {
+          const shouldMount =
+            renderState.attributesToRender.indexOf(attribute) > -1;
+
+          if (!isMounted && shouldMount) {
+            widgetsToMount.push(widget);
+            localWidgets.set(attribute, {
+              widget,
+              isMounted: true,
+            });
+          } else if (isMounted && !shouldMount) {
+            widgetsToUnmount.push(widget);
+            localWidgets.set(attribute, {
+              widget,
+              isMounted: false,
+            });
+          }
+        });
+
+        parent!.addWidgets(widgetsToMount);
+        // make sure this only happens after the regular render, otherwise it
+        // happens too quick, since render is "deferred" for the next microtask,
+        // so this needs to be a whole task later
+        setTimeout(() => parent!.removeWidgets(widgetsToUnmount), 0);
+
```

Comment:
this code could potentially be part of the widget instead of the connector since vue (and react) can rely on the host's ability to mount and unmount, but that would mean the widget itself becomes slightly more complicated, since it would have both rendering logic and widget mounting logic